File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/com/ririnto/sinon/harness/rules/ast/MutableCollectionRule.java
package com.ririnto.sinon.harness.rules.ast;

import com.ririnto.sinon.harness.rules.HarnessCheckHelper;
import com.ririnto.sinon.harness.core.RuleContext;
import com.ririnto.sinon.harness.Finding;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import com.github.javaparser.StaticJavaParser;
import com.github.javaparser.ast.CompilationUnit;
import com.github.javaparser.ast.Node;
import com.github.javaparser.ast.expr.Expression;
import com.github.javaparser.ast.expr.FieldAccessExpr;
import com.github.javaparser.ast.expr.MethodCallExpr;
import com.github.javaparser.ast.expr.ObjectCreationExpr;
import java.io.IOException;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Stream;

/**
 * Rule that forbids mutable collection instantiation.
 */
public enum MutableCollectionRule implements AstRule {
    INSTANCE;

    private static final String CATEGORY = "mutableCollection";
    private static final List<String> DEFAULT_CONSTRUCTORS = List.of("ArrayList", "HashMap", "HashSet", "LinkedList", "LinkedHashMap", "LinkedHashSet", "TreeMap", "TreeSet");
    private static final List<String> DEFAULT_FQNS = List.of("java.util.ArrayList", "java.util.HashMap", "java.util.HashSet", "java.util.LinkedList", "java.util.LinkedHashMap", "java.util.LinkedHashSet", "java.util.TreeMap", "java.util.TreeSet");
    private static final List<String> DEFAULT_ACCUMULATION_METHODS = List.of("add", "addAll", "put", "putAll");
    private static final List<String> DEFAULT_ALLOWED_BUILDERS = List.of("builder", "streamBuilder");

    @Override
    public String category() {
        return CATEGORY;
    }

    @Override
    public Collection<Finding> validate(RuleContext ctx) throws MojoExecutionException {
        final Path root = ctx.root();
        final JsonNode manifest = ctx.manifest().raw();
        final String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        try {
            final List<Path> sources = ctx.stackSources(CATEGORY);
            return sources.stream()
                    .flatMap(file -> validateMutableCollection(root, file, severity, manifest).stream())
                    .toList();
        } catch (MojoExecutionException e) {
            return List.of(Finding.of(severity, CATEGORY, "failed to enumerate sources: " + e.getMessage()));
        }
    }

    private List<Finding> validateMutableCollection(Path root, Path file, String severity, JsonNode manifest) {
        try {
            final CompilationUnit cu = StaticJavaParser.parse(file);
            final MutableConfig config = config(manifest);
            return Stream.concat(
                    cu.findAll(ObjectCreationExpr.class).stream()
                            .filter(expr -> config.isForbiddenConstructor(expr.getType().asString()))
                            .map(expr -> finding(root, file, severity, expr, expr.getType().asString())),
                    cu.findAll(MethodCallExpr.class).stream()
                            .filter(expr -> config.accumulationMethods().contains(expr.getNameAsString()))
                            .filter(expr -> expr.getScope().flatMap(MutableCollectionRule::qualifiedExpressionName).map(scope -> !config.allowedBuilders().contains(scope)).orElse(true))
                            .map(expr -> finding(root, file, severity, expr, expr.getNameAsString())))
                    .toList();
        } catch (IOException e) {
            return List.of(Finding.of(severity, CATEGORY, "failed to parse " + root.relativize(file) + ": " + e.getMessage()));
        }
    }

    private static Finding finding(Path root, Path file, String severity, Node node, String name) {
        return Finding.of(severity, CATEGORY, root.relativize(file) + ":" + node.getBegin().map(p -> p.line).orElse(-1) + ": mutable collection " + name + "; use immutable factory");
    }

    private static MutableConfig config(JsonNode manifest) {
        return new MutableConfig(
                configured(manifest, "forbiddenConstructors", DEFAULT_CONSTRUCTORS),
                configured(manifest, "forbiddenFqns", DEFAULT_FQNS),
                configured(manifest, "accumulationMethods", DEFAULT_ACCUMULATION_METHODS),
                configured(manifest, "allowedBuilders", DEFAULT_ALLOWED_BUILDERS)
        );
    }

    private static Set<String> configured(JsonNode manifest, String key, List<String> defaults) {
        final JsonNode section = manifest.get(CATEGORY);
        if (section == null) {
            return Set.copyOf(defaults);
        }
        final JsonNode parameters = section.get("parameters");
        if (parameters == null) {
            return Set.copyOf(defaults);
        }
        final List<String> configured = HarnessCheckHelper.extractPaths(parameters.get(key));
        return Set.copyOf(configured.isEmpty() ? defaults : configured);
    }

    private static Optional<String> qualifiedExpressionName(Expression expression) {
        final List<String> parts = expressionParts(expression);
        return parts.isEmpty() ? Optional.empty() : Optional.of(String.join(".", parts));
    }

    private static List<String> expressionParts(Expression expression) {
        if (expression.isNameExpr()) {
            return List.of(expression.asNameExpr().getNameAsString());
        }
        if (expression.isFieldAccessExpr()) {
            final FieldAccessExpr fieldAccess = expression.asFieldAccessExpr();
            final List<String> parts = new ArrayList<>(expressionParts(fieldAccess.getScope()));
            parts.add(fieldAccess.getNameAsString());
            return parts;
        }
        return List.of();
    }

    private record MutableConfig(Set<String> constructors, Set<String> forbiddenFqns, Set<String> accumulationMethods, Set<String> allowedBuilders) {
        private boolean isForbiddenConstructor(String typeName) {
            return constructors.contains(typeName) || forbiddenFqns.contains(typeName);
        }
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/com/ririnto/sinon/harness/rules/ast/MutableCollectionRule.kt
package com.ririnto.sinon.harness.rules.ast

import com.ririnto.sinon.harness.ast.AstFinding
import com.ririnto.sinon.harness.ast.AstSupport
import com.ririnto.sinon.harness.ast.HarnessAstResults.Finding
import com.ririnto.sinon.harness.core.RuleContext
import com.ririnto.sinon.harness.core.Severity
import com.ririnto.sinon.harness.rules.HarnessAstRule
import org.jetbrains.kotlin.com.intellij.psi.PsiElement
import org.jetbrains.kotlin.psi.KtCallExpression
import org.jetbrains.kotlin.psi.KtDotQualifiedExpression
import org.jetbrains.kotlin.psi.KtExpression
import org.jetbrains.kotlin.psi.KtFile
import org.jetbrains.kotlin.psi.KtLambdaArgument
import org.jetbrains.kotlin.psi.KtLambdaExpression
import org.jetbrains.kotlin.psi.KtNameReferenceExpression
import org.jetbrains.kotlin.psi.KtPsiFactory
import org.jetbrains.kotlin.psi.KtTreeVisitorVoid
import org.jetbrains.kotlin.psi.KtUserType
import org.jetbrains.kotlin.psi.KtValueArgument
import org.jetbrains.kotlin.psi.psiUtil.getQualifiedExpressionForSelector
import java.nio.file.Path

/**
 * Rule that delegates mutable-collection findings to source AST/PSI analysis.
 */
object MutableCollectionRule : HarnessAstRule() {
    /**
     * Category key.
     */
    override val category: String = "mutableCollection"

    override fun renderAstFindings(
        ctx: RuleContext,
        findings: Collection<AstFinding>,
    ): Collection<Finding> =
        buildList {
            findings.forEach { finding ->
                add(
                    Finding(
                        ctx.manifest.severityOf(category),
                        category,
                        "${finding.file}:${finding.line}: mutable collection `${finding.detail(
                            "name",
                        )}` is forbidden; " +
                            "use immutable alternatives",
                    ),
                )
            }
        }

    override fun findAstFindings(
        file: Path,
        ctx: RuleContext,
        astFactory: KtPsiFactory?,
    ): Collection<AstFinding> =
        buildSet {
            val ktFile = AstSupport.parse(file, astFactory)
            ktFile?.accept(
                Visitor(
                    ::add,
                    file,
                    ctx,
                    ktFile,
                    MutableConfig(
                        configuredSet(
                            ctx,
                            "forbiddenConstructors",
                            listOf(
                                "mutableListOf",
                                "mutableSetOf",
                                "mutableMapOf",
                                "arrayListOf",
                                "hashSetOf",
                                "hashMapOf",
                                "linkedSetOf",
                                "linkedMapOf",
                                "ArrayList",
                                "HashSet",
                                "HashMap",
                                "LinkedHashMap",
                                "LinkedHashSet",
                                "LinkedList",
                                "TreeMap",
                                "TreeSet",
                            ),
                        ),
                        configuredSet(
                            ctx,
                            "forbiddenTypes",
                            listOf(
                                "MutableList",
                                "MutableSet",
                                "MutableMap",
                                "ArrayList",
                                "HashSet",
                                "HashMap",
                                "LinkedHashMap",
                                "LinkedHashSet",
                                "LinkedList",
                                "TreeMap",
                                "TreeSet",
                            ),
                        ),
                        configuredSet(
                            ctx,
                            "forbiddenFqns",
                            listOf(
                                "java.util.ArrayList",
                                "java.util.HashMap",
                                "java.util.HashSet",
                                "java.util.LinkedHashMap",
                                "java.util.LinkedHashSet",
                                "java.util.LinkedList",
                                "java.util.TreeMap",
                                "java.util.TreeSet",
                                "kotlin.collections.MutableList",
                                "kotlin.collections.MutableMap",
                                "kotlin.collections.MutableSet",
                            ),
                        ),
                        configuredSet(
                            ctx,
                            "accumulationMethods",
                            listOf("add", "addAll", "put", "putAll"),
                        ),
                        configuredSet(
                            ctx,
                            "allowedBuilders",
                            listOf("buildList", "buildMap", "buildSet", "dependencies", "apply", "also"),
                        ),
                    ),
                ),
            )
        }

    private fun configuredSet(
        ctx: RuleContext,
        key: String,
        defaults: List<String>,
    ): Set<String> {
        return (ctx.manifest.stringArray(category, key).ifEmpty { defaults }).toSet()
    }

    private data class MutableConfig(
        val constructors: Set<String>,
        val types: Set<String>,
        val forbiddenFqns: Set<String>,
        val accumulationMethods: Set<String>,
        val allowedBuilders: Set<String>,
    )

    /**
     * AST visitor for mutable collection analysis.
     *
     * Identifies uses of mutable collection factories and types throughout the AST.
     */
    private class Visitor(
        private val record: (AstFinding) -> Unit,
        private val file: Path,
        private val ctx: RuleContext,
        private val ktFile: KtFile,
        private val config: MutableConfig,
    ) : KtTreeVisitorVoid() {
        override fun visitCallExpression(expression: KtCallExpression) {
            super.visitCallExpression(expression)
            val calleeName = calleeName(expression).orEmpty()
            val qualifiedParts = qualifiedCallParts(expression) ?: listOf(calleeName)
            val qualifiedName = qualifiedParts.joinToString(".")
            val receiverParts = qualifiedParts.dropLast(1)
            val isAccumulation = calleeName in config.accumulationMethods
            val findingName =
                when {
                    calleeName in config.constructors -> calleeName

                    qualifiedName in config.forbiddenFqns -> qualifiedName

                    isAccumulation && !isReceiverAllowed(receiverParts) &&
                        !isInsideAllowedBuilderLambda(expression) -> calleeName

                    else -> ""
                }
            if (findingName.isNotEmpty()) {
                recordFinding(expression, findingName)
            }
        }

        override fun visitUserType(type: KtUserType) {
            super.visitUserType(type)
            val typeParts = userTypeParts(type)
            val typeName = typeParts.lastOrNull().orEmpty()
            val qualifiedTypeName = typeParts.joinToString(".")
            val findingName = when {
                typeName in config.types -> typeName
                qualifiedTypeName in config.forbiddenFqns -> qualifiedTypeName
                else -> ""
            }
            if (findingName.isNotEmpty()) {
                record(
                    AstFinding(
                        rule = category,
                        file = AstSupport.relativeFilePath(file, ctx.root),
                        line = AstSupport.lineOf(ktFile, type.node?.startOffset),
                        details = mapOf("name" to findingName),
                    ),
                )
            }
        }

        private fun isReceiverAllowed(receiverParts: List<String>): Boolean =
            receiverParts.lastOrNull() in config.allowedBuilders

        private fun isInsideAllowedBuilderLambda(call: KtCallExpression): Boolean =
            generateSequence(call.parent as PsiElement?) { element -> element.parent }
                .filterIsInstance<KtLambdaExpression>()
                .any { lambda ->
                    val argument = lambda.parent
                    (when (argument) {
                        is KtLambdaArgument -> argument.parent as? KtCallExpression
                        is KtValueArgument -> argument.parent?.parent as? KtCallExpression
                        else -> null
                    }?.let(::calleeName).orEmpty()) in config.allowedBuilders
                }

        private fun calleeName(expression: KtCallExpression): String? =
            (expression.calleeExpression as? KtNameReferenceExpression)?.getReferencedName()

        private fun qualifiedCallParts(expression: KtCallExpression): List<String>? =
            (expression.getQualifiedExpressionForSelector() as? KtDotQualifiedExpression)?.let { qualified ->
                expressionParts(qualified.receiverExpression) + listOfNotNull(calleeName(expression))
            }

        private fun expressionParts(expression: KtExpression): List<String> =
            (when (expression) {
                is KtNameReferenceExpression -> listOf(expression.getReferencedName())
                is KtDotQualifiedExpression -> expressionParts(expression.receiverExpression) +
                    expression.selectorExpression?.let(::expressionParts).orEmpty()
                is KtCallExpression -> listOfNotNull(calleeName(expression))
                else -> emptyList()
            })

        private fun userTypeParts(type: KtUserType): List<String> =
            generateSequence(type) { userType -> userType.qualifier }
                .mapNotNull { userType -> userType.referencedName }
                .toList()
                .asReversed()

        private fun recordFinding(
            expression: KtCallExpression,
            name: String,
        ) {
            record(
                AstFinding(
                    rule = category,
                    file = AstSupport.relativeFilePath(file, ctx.root),
                    line = AstSupport.lineOf(ktFile, expression.node?.startOffset),
                    details = mapOf("name" to name),
                ),
            )
        }
    }
}
